test/shared/src/main/scala/zio/test/mock/package.scala

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2017-2020 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test
+
+import zio.{ Has, Ref, Tagged, UIO, ZLayer }
+
+package object mock {
+  type Spy = Has[Spy.Service]
+
+  /**
+   * A `Spy` wraps a live service, maintaining an internal state that is
+   * updated based on calls to that service and allowing assertions to be made
+   * based on that internal state and values returned by the live service.
+   */
+  object Spy {
+
+    trait Service {
+      def spy: UIO[TestResult]
+    }
+
+    /**
+     * Constructs a spy that wraps an existing live service. The spy maintains
+     * some internal state `S` with the specified initial value. The update
+     * function `f` takes an existing live service and returns an updated
+     * service that has access to the live service as well as a `Ref` with the
+     * internal spy state and a `Ref` to hold the result of assertions made
+     * about the internal state and values returned.
+     */
+    def live[S, A: Tagged](z: S)(
+      f: PartialFunction[(Ref[S], Ref[TestResult], Invocation[A, _, _]), UIO[Unit]]
```

Comment:
It occurs to me we can push the `Ref[S]` out of here if we want and make the user create a ref, if desired. Because basically all we're doing is passing along state the user could construct themselves before they call `live`. 

I'm not sure if this is a good idea, however. But it's more "orthogonal". 

Also: on `Ref[TestResult]`, what if we just made it return `UIO[TestResult]`? Would simplify and make easier to understand.

Reply:
Could definitely push the `Ref` out if we wanted to. Most of the heavy lifting here is done by `Spyable` that doesn't have the `Ref` for exactly that reason and you can see how it is easy to add in. Another thing we could do is have two variants, a more "fundamental" one that doesn't have the `Ref` and one implemented in terms of it for the "everyday" use case that gives you a `Ref`. That way if you wanted to do something else like use multiple `Ref`s or use a `RefM` it would be easy to do that. We could even have a couple of common baked in implementations (e.g. `Ref[S]`, `RefM[S]`).

I'm a little worried about changing the signature to return `UIO[TestResult]` because that forces the user to make an assertion about every state transition where they want to do anything, whereas in many cases they may just want to update state. For example in the prototype asserts are only being made after `get` calls, all the other calls only update state.

Reply:
Thinking about it more I am coming around to your point of view that maybe we just simplify the signature here, potentially to be the same as the one in `Spyable` and then maybe we don't even need `Spy` separate from `Spyable`. The more general `Spyable` signature already gives you the ability to maintain state by creating a `Ref` or other similar type and passing it in. This also gives the user the ability to determine how they want to make assertions.

One advantage of this is right now we have a little bit of awkwardness where the user can maintain state with the `Ref[S]` but can't actually access the `S` outside the function since we aren't parameterizing the `Spy` on `S` because services need to be monomorphic. That also creates the need for the `TestResult` because we need something monomorphic the user can always extract. If we just let the user pass in whatever they want they would have access to that state outside the function and could make whatever assertions they wanted about it.

The downside is the user potentially needs to think a little more about what to do with this function (e.g. you create some state and then you use this effect to update it) versus us kind of serving a `Ref` or two up for them.

Reply:
And then we could have more specific combinators implemented in terms of that general one. Like we could have one that takes a live service and gives you back a new service along with a `Ref` which will hold a list of all invocations that have been made on the service up to whenever you access the `Ref`, etc...

Reply:
How about `UIO[Option[TestResult]]`? Or do we have a `TestResult` the equivalent of `noOp` (looks like not).

Reply:
@adamgfraser Right!!